### PR TITLE
Infer types. Smart defaults for the visualize window. Basic implementation.

### PR DESCRIPTION
**Pull request by bkyryliuk**
Basic implementation of the type inference for smart defaults in the visualize flow for sqllab.

Tested on presto / sqlite / mysql + unit test cover postgres

Reviewers:
- @mistercrunch 
- @ascott 
- @vera-liu 

screen shot 2016-09-16 at 3 44 19 pm

### CODE REVIEW

File: caravel/assets/javascripts/SqlLab/components/ResultSet.jsx [hunk 1]
```diff
@@ -72,7 +72,7 @@ class ResultSet extends React.Component {
           <div className="ResultSet">
             <Table
               data={results.data}
-              columns={results.columns}
+              columns={results.columns.map(function (o) { return o.name; })}
```
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/components/ResultSet.jsx]: for very short function you should use the more concise arrow operator:
`(col) => coll.name`

File: caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx [hunk 2]
```diff
@@ -26,6 +26,19 @@ class VisualizeModal extends React.Component {
       columns: {},
       hints: [],
     };
+    // update columns if possible
+    this.componentWillMount();
```
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx]: I wouldn't call react's lifecycle methods directly as the framework typically handles them. Just refactor what's in `componentWillMount` into it's own method and call that one from both the constructor and `componentWillMount `

File: caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx [hunk 3]
```diff
@@ -26,6 +26,19 @@ class VisualizeModal extends React.Component {
...
+  }
+  componentWillMount() {
+    if (!this.props.query || !this.props.query.results.columns) {
+      return;
+    }
+    const columns = {};
```
Comment (mistercrunch) [inline caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx]: these 4 lines would be more concise as: 
```
this.props.query.results.columns.forEach((col) => {
   columns[col.name] = col;
});
```

File: caravel/sql_lab.py [hunk 4]
```diff
@@ -114,49 +114,123 @@ def handle_error(msg):
...
     '''
     # Hack testing using a kv store for results
     key = "query_id={}".format(query.id)
     logging.info("Storing results in key=[{}]".format(key))
     cache.set(key, json.dumps(payload, default=utils.json_iso_dttm_ser))
     '''
+
+
+THRESHOLD = 95
+SAMPLE_SIZE = 100
```
Comment (mistercrunch) [inline caravel/sql_lab.py]: constants are expected to be at the top of the module (though we should start thinking about breaking some of our large modules at this point...). As you move them to the top, you should give more specific names as in `INFER_COL_TYPES_SAMPLE_SIZE`

File: caravel/sql_lab.py [hunk 5]
```diff
@@ -114,49 +114,123 @@ def handle_error(msg):
...
+
+
+# TODO(bkyryliuk): add support for the conventions like: *_dim or dim_*
+#     - dimensions, *_ts, ts_*, ds_*, *_ds - datetime, etc.
+# TODO(bkyryliuk): recognize integer encoded enums.
+def get_columns_dict(df):
```
Comment (mistercrunch) [inline caravel/sql_lab.py]: short simple docstring on 10+ command functions, todos can be moved inside the docstring

**Comment by mistercrunch** [comment]
I'm not sure if it's in-scope for this PR (tell me what you think!), but I was thinking about where all the new methods/functions around the results should live, and I like the idea to organize all this around of a `CaravelDataFrame` that would derive `pandas.DataFrame` and add the type inference methods to it.

Eventually we can migrate more results related transformations to that class.

**Review by mistercrunch — CHANGES_REQUESTED** [review]

File: caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx (hunk 2 above)
Comment (ascott) [inline caravel/assets/javascripts/SqlLab/components/VisualizeModal.jsx]: +1

**Comment by ascott** [comment]
🎉

**Comment by mistercrunch** [comment]
LGTM

**Comment by bkyryliuk** [comment]
thanks @ascott and @mistercrunch  for the review!

**Pull request merged**